source/ast/types/AllTypes.cpp

```diff
@@ -1128,8 +1128,64 @@ const Type& VirtualInterfaceType::fromSyntax(const ASTContext& context,
     }
 
     auto loc = syntax.name.location();
-    auto& iface = InstanceSymbol::createVirtual(context, loc, def->as<DefinitionSymbol>(),
-                                                syntax.parameters);
+    auto& defSym = def->as<DefinitionSymbol>();
+    auto& iface = InstanceSymbol::createVirtual(context, loc, defSym, syntax.parameters);
+
+    // To control duplicates
+    SmallSet<const DefinitionSymbol*, 4> visited;
+    // Find cached instance hierarchical identifiers
+    std::function<std::optional<const InstanceSymbol*>(const Scope& scope)> findInst =
+        [&defSym, &findInst, &visited, &context,
+         &iface](const Scope& scope) -> std::optional<const InstanceSymbol*> {
+        for (auto& name : scope.getNameMap()) {
+            if (auto inst = name.second->as_if<InstanceSymbol>()) {
+                auto instDef = &inst->getDefinition();
+                if (instDef == &defSym && inst->body.hierIdentifiers.has_value()) {
```

Comment:
This is definitely not correct; just because the definition is the same doesn't mean the hierarchical identifiers will be the same, since behavior can vary on the parameters of the instance controlling generate blocks in the child.

This is a generally hard problem; my thinking in the past was that I would keep track of found hierarchical identifiers during Lookup and caching based on a more complicated cache key that takes into account parameters, interface ports, any active configurations, etc. Tracking for that is here: https://github.com/MikePopoloski/slang/issues/949

Reply:
Hi, I have an idea for a hash - is it can be a `std::hash` of `std::string` consisting of an instance kind (for example `interface` or `module`), instance type (for example - module type identifier) and a set of parameter values ​​with their identifiers? 

For example for such module:

```verilog
module mem_model #(
 parameter ADDR_WIDTH=8;
 parameter DATA_WIDTH=32;)
 (clk, addr, data);

 input  clk;
 input  [ADDR_WIDTH-1:0] addr;
 output [DATA_WIDTH-1:0] data;
 .....
 .....
endmodule
```

and it's instances:

```verilog
//Creates mem_1 instance with default addr and data widths.
mem_model mem_1 (.clk(clk),
                 .addr(addr_1),
                 .data(data_1));

//Creates mem_2 instance with addr width = 4 and data width = 8.
mem_model #(4,8) mem_2 (.clk(clk),
                        .addr(addr_2),
                        .data(data_2));

//Creates mem_3 instance with addr width = 32 and data width = 64.
mem_model #(32,64) mem_3 (.clk(clk),
                          .addr(addr_3),
                          .data(data_3));

//Creates mem_4 instance with default addr width and data width = 64.
mem_model #(DATA_WIDTH=64) mem_4 (.clk(clk),
                                 .addr(addr_3),
                                 .data(data_3));
```

hash string can be:

```с++
"module mem_model(ADDR_WIDTH=8, DATA_WIDTH=32)" // for mem_1

"module mem_model(ADDR_WIDTH=4, DATA_WIDTH=8)" // for mem_2

"module mem_model(ADDR_WIDTH=32, DATA_WIDTH=64)" // for mem_3

"module mem_model(ADDR_WIDTH=8, DATA_WIDTH=64)" // for mem_4
```

What do you think about this suggestion?

Reply:
There's no reason for it to be a string; you can just hash the relevant fields. This diff from a long time ago removed an InstanceCache that I had in slang at one point: https://github.com/MikePopoloski/slang/commit/0892da30760454a58bdfbb9a2d1d40d728643e36#diff-18c21a73a1abf2b9e1aa72758848d9bceee11a575bfab7d762dfc51217ec26c3

That's a good start but it also needs to work correctly with hierarchical references into and out of cached instances, which is very tricky and something I need to think more carefully about (which is one of the reasons why I removed this back in the day).